File: impilo/src/main/java/org/fhi360/plugins/impilo/domain/entities/Patient.java
package org.fhi360.plugins.impilo.domain.entities;

import com.blazebit.persistence.view.*;
import io.github.jbella.snl.core.api.domain.Organisation;
import io.github.jbella.snl.core.api.id.UUIDV7;
import jakarta.persistence.Entity;
import jakarta.persistence.Id;
import jakarta.persistence.ManyToOne;
import jakarta.validation.constraints.NotNull;
import lombok.Getter;
import lombok.Setter;
import org.fhi360.plugins.impilo.domain.providers.*;

import java.time.LocalDate;
import java.util.List;
import java.util.UUID;

@Entity
@Getter
@Setter
public class Patient {
    @Id
    @UUIDV7
    private UUID id;

    private String givenName;

    private String familyName;

    private LocalDate dateOfBirth;

    private String sex;

    private String phoneNumber;

    private String address;

    private String hospitalNumber;

    private String regimen;

    private LocalDate nextAppointmentDate;

    private String personId;

    private String patientId;

    private String facilityId;

    private String facilityName;

    private UUID reference;

    @ManyToOne
    @NotNull
    private Organisation organisation;

    @EntityView(Patient.class)
    public interface IdView {
        @IdMapping
        UUID getId();
    }

    @EntityView(Patient.class)
    @CreatableEntityView
    public interface CreateView extends IdView {
        String getFamilyName();

        void setFamilyName(String name);

        String getGivenName();

        void setGivenName(String name);

        LocalDate getDateOfBirth();

        void setDateOfBirth(LocalDate dateOfBirth);

        String getSex();

        void setSex(String sex);

        String getPhoneNumber();

        void setPhoneNumber(String phoneNumber);

        String getAddress();

        void setAddress(String address);

        String getHospitalNumber();

        void setHospitalNumber(String hospitalNumber);

        String getRegimen();

        void setRegimen(String regimen);

        LocalDate getNextAppointmentDate();

        void setNextAppointmentDate(LocalDate appointmentDate);

        String getPatientId();

        void setPatientId(String patientId);

        String getPersonId();

        void setPersonId(String personId);

        String getFacilityId();

        void setFacilityId(String facilityId);

        String getFacilityName();

        void setFacilityName(String facilityName);

        UUID getReference();

        void setReference(UUID reference);

        Organisation.IdView getOrganisation();

        void setOrganisation(Organisation.IdView organisation);
    }

    @EntityView(Patient.class)
    @UpdatableEntityView
    public interface UpdateView extends CreateView {
        void setId(UUID id);
    }

    @EntityView(Patient.class)
    public record ListView(
        @IdMapping UUID id, String givenName, String familyName, String sex,
        LocalDate dateOfBirth, String address,
        String phoneNumber, String hospitalNumber, String regimen,
        String facilityName,
        @MappingSubquery(PatientSiteNameSubqueryProvider.class)
        String site
    ) {
    }

    @EntityView(Patient.class)
    public record View(@IdMapping UUID id, String givenName, String familyName, String sex,
                       LocalDate dateOfBirth, String address, LocalDate nextAppointmentDate,
                       String phoneNumber, String hospitalNumber, @Mapping("regimen") String assignedRegimen,
                       @MappingSubquery(LastRefillDateSubqueryProvider.class)
                       LocalDate lastRefillDate,
                       @MappingSubquery(LastClinicDateSubqueryProvider.class)
                       LocalDate lastClinicVisit,
                       @MappingSubquery(DateOfNextRefillSubqueryProvider.class)
                       LocalDate nextRefillDate,
                       @MappingSubquery(PatientSiteCodeSubqueryProvider.class)
                       UUID siteCode,
                       @MappingSubquery(PatientSiteNameSubqueryProvider.class)
                       String facility) {
    }
}

File: impilo-mobile/lib/backend/floor/entities/patient.dart
import 'package:floor/floor.dart';

@entity
class Patient {
  @PrimaryKey(autoGenerate: true)
  int? id;
  String givenName;
  String familyName;
  String hospitalNo;
  String uniqueId;
  DateTime dateOfBirth;
  String sex;
  String phoneNumber;
  String assignedRegimen;
  String facility;
  String siteCode;
  String address;
  DateTime lastClinicVisit;
  DateTime lastRefillDate;
  DateTime nextAppointmentDate;
  DateTime nextRefillDate;
  bool serviceDiscontinued;
  String reasonDiscontinued;
  DateTime dateDiscontinued;
  String uuid;
  bool synced;

  Patient(
      this.id,
      this.givenName,
      this.familyName,
      this.hospitalNo,
      this.uniqueId,
      this.dateOfBirth,
      this.sex,
      this.phoneNumber,
      this.assignedRegimen,
      this.facility,
      this.siteCode,
      this.address,
      this.lastClinicVisit,
      this.lastRefillDate,
      this.nextAppointmentDate,
      this.nextRefillDate,
      this.serviceDiscontinued,
      this.reasonDiscontinued,
      this.dateDiscontinued,
      this.uuid,
      this.synced);

  factory Patient.fromJson(Map<String, dynamic> row) => Patient(
      null,
      row['givenName'],
      row['familyName'],
      row['hospitalNumber'],
      row['uniqueId'] ?? '',
      DateTime.parse(row['dateOfBirth']),
      row['sex'],
      row['phoneNumber'],
      row['assignedRegimen'],
      row['facility'],
      row['siteCode'],
      row['address'],
      DateTime.tryParse(row['lastClinicVisit'] ?? '') ?? DateTime(1900),
      DateTime.tryParse(row['lastRefillDate'] ?? '') ?? DateTime(1900),
      DateTime.tryParse(row['nextAppointmentDate'] ?? '') ?? DateTime(1900),
      DateTime.tryParse(row['nextRefillDate'] ?? '') ?? DateTime(1900),
      false,
      '',
      DateTime(1970),
      row['id'],
      true);

  Map<String, dynamic> toJson() => {
        'id': id,
        'givenName': givenName,
        'familyName': familyName,
        'hospitalNo': hospitalNo,
        'uniqueId': uniqueId,
        'dateOfBirth': dateOfBirth.toIso8601String(),
        'sex': sex,
        'phoneNumber': phoneNumber,
        'assignedRegimen': assignedRegimen,
        'facility': facility,
        'siteCode': siteCode,
        'address': address,
        'lastClinicVisit': lastClinicVisit.toIso8601String(),
        'lastRefillDate': lastRefillDate.toIso8601String(),
        'serviceDiscontinued': serviceDiscontinued,
        'reasonDiscontinued': reasonDiscontinued,
        'dateDiscontinued': dateDiscontinued.toIso8601String()
      };
}
